finish HW

vlad-bb opened the pull request:
Try to implement syncronization only on levels Collection and Store, because UserService use Collection and Store which already sync. Maybe need to syncronize UserService also?
But I am not sure about this.

vlad-bb committed:
finish HW

Try to implement syncronization only on levels Collection and Store, because UserService use Collection and Store which already sync.
Maybe need to syncronize UserService also?
But I am not sure about this.

stashkevi4 commented on hw_11/internal/document_store/collection.go:
```diff
@@ -0,0 +1,322 @@
+		sortedKeys: make([]string, 0),
+	}
+
+	for key, doc := range c.documents {
```
On range through docs RLock also needed

stashkevi4 commented on hw_11/internal/document_store/collection.go:
```diff
@@ -0,0 +1,322 @@
+		index.values[strValue][key] = struct{}{}
+	}
+
+	c.indexes[fieldName] = index
```
indexes is also map which requires its own mutex

stashkevi4 commented on hw_11/internal/document_store/collection.go:
```diff
@@ -0,0 +1,322 @@
+
+type Index struct {
+	fieldName  string
+	values     map[string]map[string]struct{}
+	sortedKeys []string
```
These fields also require mutexes to be thread-safe

vlad-bb committed:
fix mutex

add mutex
now if I run like `go run -race cmd/app/main.go`
IDE  no raise any errors, so that code thread safe)

stashkevi4 commented on hw_11/internal/document_store/collection.go:
```diff
@@ -173,33 +183,40 @@ func (c *Collection) CreateIndex(fieldName string) error {
 			continue
 		}
-
+		index.mu.Lock()
```
index was just created by current goroutine, other goroutines doesn't have access to it anyway, it's safe to use it without mutexes.

Mutexes should be protecting only shared resources

stashkevi4 commented on hw_11/internal/document_store/collection.go:
```diff
@@ -0,0 +1,337 @@
+
+type Collection struct {
+	cfg       CollectionConfig
+	documents map[string]Document
+	indexes   map[string]*Index
```
For production ready solution you'd need to have separate mutex per shared resource. They are independent and when index is traversed it doesn't mean that document couldn't be added
